### PR TITLE
Auction system for the registar

### PR DESCRIPTION
**Pull request by chriseth**
This only compiles but is not tested. Comments are welcome.

### CODE REVIEW

**Comment by alexvandesande** [comment]
Can you explain what you're trying to do? It's my understanding that this is an open ended auction model where the winner only pays the second highest bid. As far as I understood also, the previous owner gets 1% of the money raised on the auction, but the other bidders never get anything back. Is that right?

**Comment by chriseth** [comment]
I tried to implement what was written as a "TODO"-comment in the previous version.
The mechanics - which are just a suggestion, I have no idea whether they will work out - are as follows:

The bidding time is 7 days and is extended by another 7 days from the point on where a higher bid is received.
If the bidding time ended without a new bid, the address from which the highest bid came is the winner.
The winner receives the item and, 
if the item had a previous owner: the previous owner receives all bids minus 1 percent of the highest bid. 
If it did not have a previous owner: the winner receives the difference between the highest and the second highest bid, all other bids are burned.

The burning of the bids is supposed to be a countermeasure against always extending the bidding time by adding one wei to the highest bid and thus delaying the outcome of the auction. 
Furthermore, the fact that the previous owner gets almost all bids allows her to keep a name at almost no cost.

**Comment by alexvandesande** [comment]
I’m afraid that this particular system is extremely wasteful and will end up benefiting DAOs with large amounts of resources that can threaten to bid irrationally high (but seldom needing to do it) which will then charge rent for smaller entities. Basically feudal lords of the blockchain :-/

Instead of having to decide this now, I’d say the most important feature that the name registrar should have would be to allow domain name holders to be able to vote to switch or update the system. Basically have a vote to where the “registrar” name is pointed at.

This way we can allow the whole community to have these debates whenever a problem arises

File: GlobalRegistrar/contract.sol [hunk 1]
```diff
@@ -1,25 +1,65 @@
 //sol
 
-import "owned";
-
 contract NameRegister {
-	function addr(bytes32 _name) constant returns (address o_owner) {}
-	function name(address _owner) constant returns (bytes32 o_name) {}
+	function addr(bytes32 _name) constant returns (address o_owner);
+	function name(address _owner) constant returns (bytes32 o_name);
```
Comment (CJentzsch) [inline GlobalRegistrar/contract.sol]: What about one owner having multiple names?

File: GlobalRegistrar/contract.sol [hunk 2]
```diff
@@ -1,25 +1,65 @@
...
 contract Registrar is NameRegister {
 	event Changed(bytes32 indexed name);
 	event PrimaryChanged(bytes32 indexed name, address indexed addr);
 
-	function owner(bytes32 _name) constant returns (address o_owner) {}
-	function addr(bytes32 _name) constant returns (address o_address) {}
-	function subRegistrar(bytes32 _name) constant returns (address o_subRegistrar) {}
-	function content(bytes32 _name) constant returns (bytes32 o_content) {}
+	function owner(bytes32 _name) constant returns (address o_owner);
+	function addr(bytes32 _name) constant returns (address o_address);
```
Comment (CJentzsch) [inline GlobalRegistrar/contract.sol]: This is confusing. You have defined "function addr(bytes32 _name) constant returns (address o_owner);" in contract NameRegister, and are here redefining it with the exact same interface but different return value. Can you explain that to me?

File: GlobalRegistrar/contract.sol [hunk 3]
```diff
@@ -1,25 +1,65 @@
...
+	function subRegistrar(bytes32 _name) constant returns (address o_subRegistrar);
+	function content(bytes32 _name) constant returns (bytes32 o_content);
 	
-	function name(address _owner) constant returns (bytes32 o_name) {}
+	function name(address _owner) constant returns (bytes32 o_name);
+}
+
+contract AuctionSystem {
```
Comment (CJentzsch) [inline GlobalRegistrar/contract.sol]: Another question about separating the contracts. In the contract AuctionSystem is no function to actually bid on a name. Doesn't it belong here?
Also in the NameRegister contract I can see no data members, but isn't that the place for them? I think if you can not seperate it like that, you should put it all in one contract. For me separation makes only sense if you can reuse the basic contracts and/or can use them as they are, with other derived contracts only adding features to working stand-alone contracts.

File: GlobalRegistrar/contract.sol (hunk 1 above)
Comment (chriseth) [inline GlobalRegistrar/contract.sol]: Let us address this in a second PR.

File: GlobalRegistrar/contract.sol (hunk 3 above)
Comment (chriseth) [inline GlobalRegistrar/contract.sol]: I guess Registrar and NameRegister are meant to be plain interfaces (one to retrieve names and one to register them), but I actually do not know what the intention of the original author was. We can address this in a second PR.

Concerning the bidding function, I will try to clean it up.

**Comment by jmiehau** [comment]
I was thinking for a strategy but I don´t know if we can implementing on the protocol:

 - Increment exponentially the fee to renovate if you have not been using the "name" on that period for the actual owner. And, It can be chance an auction to bid for other users for the next period. The owner to participate in the auction should pay more than the others (If he use the same address).

- In case you have use it for your Dapps, attending of the volume of transactions and contracts related, the renewal fee for the owner will decrease exponentially at the time it increase exponentially for the others. To protect the owner´s registration and give him stability.

**Comment by chriseth** [comment]
Tracking something like "usage" inside a contract is very hard to do and even if it is possible, I guess it would be easy to fake.

**Comment by jmiehau** [comment]
And how about tracking the Gas consumption? If you used the network through transaction you have payed fees. It can be a way the calculate the usage of the "name" involved.

**Comment by alexvandesande** [comment]
Usage or gas consumption means nothing about the value of a name. The value
should be simply based on demand: the current owner should pay a renewal
fee which is proportional to the amount other people are willing to bid to
have it.

On Fri, Jun 12, 2015 at 8:08 AM, Jorge Mielgo <notifications@github.com>
wrote:

> And how about tracking the Gas consumption? If you used the network
> through transaction you have payed fees. It can be a way the calculate the
> usage of the "name" involved.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ethereum/dapp-bin/pull/17#issuecomment-111453744>.
>



-- 

Alex Van de Sande
UX Designer

**Pull request merged**